runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -325,20 +326,23 @@ public void syncObjectUnsafe(long timestamp) {
                 this.optimisticStream = null;
             }
             // If we are too far ahead, roll back to the past
-            if (getVersionUnsafe() > timestamp) {
-                try {
-                    rollbackObjectUnsafe(timestamp);
-                    // Rollback successfully got us to the right
-                    // version, we're done.
-                    if (getVersionUnsafe() == timestamp) {
-                        return;
-                    }
-                } catch (NoRollbackException nre) {
-                    log.warn("Rollback[{}] to {} failed {}", this, timestamp, nre);
-                    resetUnsafe();
-                }
-            }
+            this.rollBackToTimestamp(timestamp);
+            // Attempt to sync stream to timestamp, it might have to sync to the tail to build the
+            // history
             syncStreamUnsafe(smrStream, timestamp);
+            // Check if stream was synced ahead and roll back to timestamp if necessary
+            this.rollBackToTimestamp(timestamp);
```

Comment:
the object is rolled back, then synced forward. Why do you need an extra roll back after the sync ?